### ISSUE TITLE
HelperSabotage shows numeric value 7 in HASS

### ISSUE THREAD

**Issue by dagobert** [danielperna84/pyhomematic#221]
**Describe the bug**

I have just added a HM-SEC-SC-2 and tested Sabotage alarm. I noticed that the value 7 is indicating sabotage. In home assistant I can see the value 7 but shouldn't it be `Yes` instead?
When there is no sabotage alarm present, the value in HASS is correct `No`.

https://github.com/danielperna84/pyhomematic/blob/e298897c455340cd43c50e700d227322e4ed735f/pyhomematic/devicetypes/helper.py#L16

**Comment by danielperna84** [comment]
This is cause by Home Assistant [here](https://github.com/home-assistant/home-assistant/blob/13421b326bcd25def93cd6c6dcf12a8d22203e6e/homeassistant/components/homematic/__init__.py#L112). I don't know why it has been done this way, but I assume different devices use different numbers. `0` always is _no error_, whereas anything greater than `0` is something the user has to figure out. There can be different errors after all I guess. Just saying `Yes` won't help the user to track down the issue.

**Comment by dagobert** [comment]
Are you sure. For me it looks like that ERROR in the code you referenced is always related / translated to sabotage. Hence, at least in the homematic component can never mean anything else. Also I checkt the comprehensive Homematic device description of Homegear ant ERROR is always meaning sabotage alarm.

Last, the homematic component in HASS is using your python lib. Therefor it is actually you who defines what ERROR means. the component in HASS only translates your values into readable words.

**Comment by danielperna84** [comment]
> Therefor it is actually you who defines what ERROR means.

No. pyhomematic is just a wrapper that passes on data. If the CCU sends the value 7, then we pass that on to Home Assistant. What the error code 7 means ahs nothing to do with pyhomematic. The mapping in HASS just know that with 0 everything is fine, 1 is an error, and everything else it doesn't know about.

The only reference for what `ERROR` means I found in the XML-RPC API documentation (5.1.1.4 ERROR):

```
{1, “Unknown Error”}
{2, “Busy”}
{3, “MemFull”}
{4, “Target Invalid”}
{5, “Channel Invalid”}
```

So everything, like 7, is outside of the standard errors and, in my opinion, depends on the device. So, yeah, `HelperSabotage` should be `HelperError`, and the sabotage-method should compare the value to 7. At least for RF devices. For IP devices it might be different.

**Comment by dagobert** [comment]
I am sorry about my too overconfident sentence that ERROR always means sabotage. I checkt many devices but not all. I was too fast with this.

> And `HM-Sec-SC-2` has:
> 
> * 0 = NO_ERROR
> * 1 = SABOTAGE

Here the documentation is wrong I have the idea that the document is not up-to-date anyway. As I said, I have the sensor just added with the latest firmware and it shows value "7" when it detects a sabotage. I have the idea that the documentation of [Homegear about the devices](https://ref.homegear.eu/device.html?directory=HomeMatic+BidCoS&file=rf_sc.xml&familyLink=homematicbidcos&name=HM-Sec-SC-2) is much more reliable.

There is still one thing I do not understand. You are casting the value to bool (True or False) but still the actual value is passed to HASS. When I cast "7" to bool than I get true. How does that work? Does this mean HASS is responsible to know about homematic specific device attributes? This would mean that ever project using your lib would have to figure out themselves.

If You designed your lib to be "just" a connector without handling any knowledge about the devices I understand this. Don't you thing that it could make sense to make the lib aware of  device specifics - at least in these corner cases. For example translate the information about the ERROR if known and create an attribute SABOTAGE even though it does not exist. ERROR will still be offered untouched. SABOTAGE or ERROR_SABOTAGE exists for anyways on other devices. Or instead of passing the value you could pass the translated string such as VALVE_DRIVE_BLOCKED instead of 1.

**Comment by danielperna84** [comment]
> Here the documentation is wrong I have the idea that the document is not up-to-date anyway. As I said, I have the sensor just added with the latest firmware and it shows value "7" when it detects a sabotage. I have the idea that the documentation of [Homegear about the devices](https://ref.homegear.eu/device.html?directory=HomeMatic+BidCoS&file=rf_sc.xml&familyLink=homematicbidcos&name=HM-Sec-SC-2) is much more reliable.

And that's exactly why I __don't__ want pyhomematic to take care of this. If this was part of pyhomematic, this would have to be maintained for every device with every firmware update. That's just not worth the effort. So we just pass on the value. And if a user want's to know what it is he can google it. It's different for every device after all. Do you know ioBroker? It does even less logic than pyhomematic. It just exposes all channels with all paramters with their values. The user has to decide what to do with these values. That's what I can live with for everything that's not _core_ functionality for the devices.

> There is still one thing I do not understand. You are casting the value to bool (True or False) but still the actual value is passed to HASS. When I cast "7" to bool than I get true. How does that work?

In Python everything that is not 0 or `None` will evaluate to `True`.

> Does this mean HASS is responsible to know about homematic specific device attributes? This would mean that ever project using your lib would have to figure out themselves.

Correct. I'm not willing to clean up the mess ELV has made. In theory we could detect all this by iterating through the paramsets. By doing that we wouldn't even have to create alle these classes. Everything could be automated. But reading paramsets is incredibly slow. So in some way what pyhomematic does is making usable objects out of the paramsets, but manually by us, because in terms of speed it's not practical to automate it.

> If You designed your lib to be "just" a connector without handling any knowledge about the devices I understand this. Don't you thing that it could make sense to make the lib aware of device specifics - at least in these corner cases.

I want to stick to the standards. Creating a _virtual_ `SABOTAGE` parameter that just is computed by pyhomematic would fall out of the specification. Let's just say some other tool would use pyhomematic and it didn't know about the `SABOTAGE` paramter because it does not exist for that particular device. Maybe the fired event would have negative impacts in this case.

To sum this up. I want to keep pyhomematic as simple as possible. It's a wrapper with a few extras. Everything beyond that has to be done by the consuming library if it's really needed.

**Comment by dagobert** [comment]
> To sum this up. I want to keep pyhomematic as simple as possible. It's a wrapper with a few extras. Everything beyond that has to be done by the consuming library if it's really needed.

I understand, also because it would be a lot of effort if we would expect the Attributes to stay the same across firmware updates. I personally think it would be worth it but only when there would be a active community that reports the details as they change. That is probably wishful thinking.
In addition your other points make sense from the perspective you explained.

I guess than we have to change the Helper later and contact HASS about their problematic line of code you referenced.

**Comment by danielperna84** [comment]
> I guess than we have to change the Helper later and contact HASS about their problematic line of code you referenced.

I'm the one maintaining it there. This library primarily is for use with Home Assistant. That's what I have created it for. So I'll take care of that. 😉
